### ISSUE TITLE
Object: metadata.generation can become > state.observedGeneration and rook does not set observedGeneration == generation

### ISSUE THREAD

**Issue by jhoblitt** [rook/rook#13799]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Metadata only changes to various CRs. E.g. `CephObjectStore`, `CephNFS`, etc. can result `metadata.generation` being incremented and the operator, while running, will not update `state.observedGeneration` to be in indicate that the CR is "in sync".  Restarting the operator pod will cause `state.observedGeneration` to be updated.  However, I suspect that if the delta between `observedGeneration` and `generation` is > 1 it may take multiple operator restart and settling cycles to get the values back into sync.

**Expected behavior:**

I expect that `metadata.generation` == `state.observedGeneration` when the resource is "in sync" and there is no action to be taken by an operator.  It turns out that this expectation is baked into at least fleet: **https://github.com/rancher/fleet/blob/3f0b61ff8d1d5c06f12b1c8da0dc657abb0bffc0/internal/cmd/controller/reconciler/bundle_controller.go#L74**

**How to reproduce it (minimal and precise):**

Create a `CephObjectStore` (or most other resource) and edit the annotations/labels multiple times before the operator and converged the resource.

Example of a CR that is out of sync due to annotation/label edits:

```yaml
apiVersion: ceph.rook.io/v1
kind: CephFilesystem
metadata:
  annotations:
    meta.helm.sh/release-name: rook-ceph-conf
    meta.helm.sh/release-namespace: rook-ceph
    objectset.rio.cattle.io/id: default-kueyen-fleet-lib-rook-ceph-conf
  creationTimestamp: "2023-07-25T17:25:59Z"
  finalizers:
  - cephfilesystem.ceph.rook.io
  generation: 5
  labels:
    app.kubernetes.io/managed-by: Helm
    objectset.rio.cattle.io/hash: 458995423cf8d71dceddccede394bee5173ba221
  name: obs-env
  namespace: rook-ceph
  resourceVersion: "112080565"
  uid: 66bd2b82-23e1-464b-a01b-7eb69d1c6fa8
spec:
  dataPools:
  - erasureCoded:
      codingChunks: 0
      dataChunks: 0
    failureDomain: host
    mirroring: {}
    quotas:
      maxSize: 2Ti
    replicated:
      size: 3
    statusCheck:
      mirror: {}
  metadataPool:
    erasureCoded:
      codingChunks: 0
      dataChunks: 0
    failureDomain: host
    mirroring: {}
    quotas:
      maxSize: 10Gi
    replicated:
      size: 3
    statusCheck:
      mirror: {}
  metadataServer:
    activeCount: 3
    activeStandby: true
    placement: {}
    resources:
      limits:
        cpu: "4"
        memory: 4Gi
      requests:
        cpu: "4"
        memory: 4Gi
  statusCheck:
    mirror: {}
status:
  observedGeneration: 4
  phase: Ready
```
**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Environment**:
* OS (e.g. from /etc/os-release): EL9
* Kernel (e.g. `uname -a`): ` 5.14.0-362.18.1.el9_3.x86_64`
* Rook version (use `rook version` inside of a Rook Pod): `1.11.2`
* Storage backend version (e.g. for ceph do `ceph -v`): `ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)`
* Kubernetes version (use `kubectl version`): `Server Version: v1.25.9`
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): rke[1]
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): `HEALTH_OK`

**Comment by travisn** [comment]
Per discussion in the community meeting, sounds like the predicate is canceling the reconcile, but we should consider allowing the reconcile to continue if the generation is not current. However, this may also cause unnecessary reconciles for metadata/annotation updates. @BlaineEXE thoughts?

**Comment by BlaineEXE** [comment]
Thanks for capturing this.

IMO, using generation-vs-observed as the main predicate filter is probably the overall best thing we can do. The two reasons are:
1. This is probably annoying enough to users to be worth fixing, and
2. I think it can be reduced to a low (less than 1 on average) likelihood of happening IRL

#### RE: point 1

@jhoblitt is a bit of an exceptional user by coming to us with an expression of confusion and openness to how it is resolved. In comparison, I think a lot of other users would be more immediately frustrated by this behavior, making it something that has value to fix.

#### RE: point 2

I suspect that one of the additional generations is due to the finalizer that Rook puts on the CephObjectStore. To resolve this, Rook can re-Get the CephObjectStore after applying the finalizer (waiting in a loop to make sure the version we get has the finalizer) to eliminate 1 additional reconcile caused by Rook itself.

I'm less certain about what happens to the generation when the status is first initialized (right after the finalizer). It could add a generation, but I doubt this based on k8s specs. To be safe, we could change the initial reconcile flow to:
1. add the finalizer
2. init the status
3. re-Get the latest-generation COS
4. proceed with the rest of the reconcile

I looked into the reconcile code to see if we can use the generation-observed predicate and still skip the reconcile once in the loop, but I don't think it's possible. Once we get to the reconcile, we don't have old-vs-new object context to be able to make the decision about whether we can skip the reconcile. This likely means that we need to consider the impact of extra reconciles.

Controllers like Fleet can add annotations/etc. pretty quickly, and I suspect that additional annotations/etc. may already be added by the time we re-Get the COS post-finalizer, meaning there could be cases where there are zero additional reconciles. In the case where Fleet isn't as quick to update, there would likely only be one additional reconcile. 

Assessing the impact of this for users long-term, I doubt that the extra reconciles will happen very often: probably only when the CephObjectStore is modified (and possibly after Rook version upgrades). Appropriately cautious users are likely to make these kinds of changes during business quiet hours, limiting the impact of the additional reconcile in their clusters. We also know that there are very few users who have more than 8 object stores in their clusters, meaning very few users are likely to even notice 1 additional reconcile per-store.

**Comment by BlaineEXE** [comment]
Got it. Thanks :)

Let's use my previous comment as a starting point for 2 PR fixes then:
1. eliminate the extra reconcile caused by adding the finalizer and status initialization
2. modify the predicate to use generation vs observed generation instead of spec diff

**Comment by raaizik** [comment]
Hey @jhoblitt, are you actively working on this?

**Comment by jhoblitt** [comment]
> Hey [@jhoblitt](https://github.com/jhoblitt), are you actively working on this?

No. There are currently several features higher on my priority list.

**Comment by BlaineEXE** [comment]
@raaizik messaged me offline to ask about picking this issue up, so I've assigned it. Feel free to reach out to me with any questions about contributing to Rook or working on the tasks for this issue.
